### PR TITLE
Vinai/alter drop pks

### PR DESCRIPTION
**Pull request by VinaiRachakonda**

### CODE REVIEW

File: go/libraries/doltcore/merge/merge_schema.go [hunk 1]
```diff
@@ -28,6 +28,7 @@ type conflictKind byte
 const (
 	TagCollision conflictKind = iota
 	NameCollision
+	PkCollision
```
Comment (VinaiRachakonda) [inline go/libraries/doltcore/merge/merge_schema.go:31]: Note to self: Come back to this

File: go/libraries/doltcore/diff/table_deltas.go [hunk 2]
```diff
@@ -349,6 +370,76 @@ func getKeylessDeltas(ctx context.Context, fromRoot, toRoot *doltdb.RootValue) (
...
+		return nil, nil, err
+	}
+
+	err = toRoot.IterTables(ctx, func(name string, tbl *doltdb.Table, sch schema.Schema) (stop bool, err error) {
+		delta := TableDelta{
+			ToName:  name,
+			ToTable: tbl,
+		}
+
+		// TODO: is this the most efficient way to to do this
```
Comment (VinaiRachakonda) [inline go/libraries/doltcore/diff/table_deltas.go:401]: Note to self: Come back to this

File: go/libraries/doltcore/schema/column.go [hunk 3]
```diff
@@ -149,8 +149,7 @@ func (c Column) Equals(other Column) bool {
 // kinds of values at the storage layer, but may have different constraints or type parameters.
 func (c Column) Compatible(other Column) bool {
 	return c.Tag == other.Tag &&
-		c.Kind == other.Kind &&
-		c.IsPartOfPK == other.IsPartOfPK
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/schema/column.go:153]: This is suspect, what's the reasoning for this?

File: go/libraries/doltcore/sqle/dtables/diff_table.go [hunk 4]
```diff
@@ -368,6 +373,11 @@ func (dp diffPartition) getRowIter(ctx *sql.Context, ddb *doltdb.DoltDB, ss *sch
 	toCmInfo := commitInfo{types.String(dp.toName), dp.toDate, toCol.Tag, toDateCol.Tag}
 
 	rd := diff.NewRowDiffer(ctx, fromSch, toSch, 1024)
+	if _, ok := rd.(*diff.EmptyRowDiffer); ok {
+		ctx.Warn(PrimaryKeyChanceWarningCode, fmt.Sprintf(PrimaryKeyChangeWarning, dp.fromName, dp.toName))
+		return sql.RowsToRowIter(), nil
+	}
+
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/sqle/dtables/diff_table.go:380]: Why do we check this here? Where do we return `EmptyRowDIffer`?

File: go/libraries/doltcore/sqle/dtables/diff_table.go [hunk 5]
```diff
@@ -551,6 +579,23 @@ func (dp *diffPartitions) Next() (sql.Partition, error) {
 		}
 
 		if next != nil {
+			// We want to render the dolt_diff table from a top down commit log traversal until we reach a point
+			// where we cannot diff two commits due to a schema difference. This logic is similar to whats in
+			// dp.getRowIter() for two reasons.
+			// 1. The dolt_commit_diff table exclusively uses getRowIter and not the diffPartitions object
+			// 2. The GMS partition exchange does not process partitions in the order they are sent in. This can allow
+			// commits with the computable diffs before the schema change to be rendered, convoluting the UX of the diff
+			// table.
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/sqle/dtables/diff_table.go:588]: It's a good idea to document this facet of the diff table, but this comment is too implementation-oriented and won't make sense to the next person reading this code. See comment suggestions.

Comment (andrew-wm-arthur) [reply go/libraries/doltcore/sqle/dtables/diff_table.go:588]:
```suggestion
// if we can't diff this commit with its parent, don't traverse any lower
```

File: go/libraries/doltcore/sqle/dtables/diff_table.go [hunk 6]
```diff
@@ -383,6 +393,24 @@ func (dp diffPartition) getRowIter(ctx *sql.Context, ddb *doltdb.DoltDB, ss *sch
 	}, nil
 }
 
+func (dp *diffPartition) isDiffablePartition(ctx *sql.Context) (bool, error) {
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/sqle/dtables/diff_table.go:396]:
```suggestion
// isDiffablePartition checks if the commit pair for this partition is "diffable".
// If the primary key sets changed between the two commits, it may not be
// possible to diff them.
func (dp *diffPartition) isDiffablePartition(ctx *sql.Context) (bool, error) {
```

File: go/libraries/doltcore/sqle/tables.go [hunk 7]
```diff
@@ -1720,3 +1722,255 @@ func (t *AlterableDoltTable) constraintNameExists(ctx *sql.Context, name string)
 
 	return false, nil
 }
+
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/sqle/tables.go:1725]: the rest of the schema change logic is located in `schema/alterschema` we should move this logic there and call to it

Comment (andrew-wm-arthur) [reply go/libraries/doltcore/sqle/tables.go:1725]: We should also add some golang integration test for rewriting primary and secondary indexes

File: integration-tests/bats/schema-changes.bats [hunk 8]
```diff
@@ -254,3 +254,452 @@ SQL
...
+
+    dolt sql -q "INSERT INTO t values (2,2,2)"
+    dolt commit -am "cm3"
+
+    run dolt merge test
+    [ "$status" -eq 1 ]
+    [[ "$output" =~ 'different primary key definitions for our column pk and their column pk' ]] || false
+}
+
+@test "schema-changes: same as the previous test but in the opposite direction" {
```
Comment (andrew-wm-arthur) [inline integration-tests/bats/schema-changes.bats:502]: needs a more descriptive title, tests can be inserted between these two

File: integration-tests/bats/schema-changes.bats [hunk 9]
```diff
@@ -254,3 +254,452 @@ SQL
     [[ "$output" =~ '>   `v2` VARCHAR(1024)' ]] || false
     [[ "$output" =~ 'PRIMARY KEY' ]] || false
 }
+
```
Comment (andrew-wm-arthur) [inline integration-tests/bats/schema-changes.bats:257]: nit: great test coverage here, but there are so many new ones, you should split them out into their own file 
`primary-key-change.bats`

**Review by andrew-wm-arthur — COMMENTED** [review]
Generally looks good. 
This covers the major dimensions of the product surface area and has thorough test coverage.
We should discuss what this means for SQL diff and import, and see what's worth pursuing there.

File: go/libraries/doltcore/schema/column.go (hunk 3 above)
Comment (VinaiRachakonda) [reply go/libraries/doltcore/schema/column.go:153]: The superschema will complain about a tag collision otherwise during. [cc:](https://github.com/dolthub/dolt/blob/83b4aaaf7413f9e49ff8c6c037b12888fe8b95c1/go/libraries/doltcore/schema/super_schema.go#L67)

I also don't think it's correct anymore since primary key changes are possible.

File: go/libraries/doltcore/sqle/dtables/diff_table.go (hunk 4 above)
Comment (VinaiRachakonda) [reply go/libraries/doltcore/sqle/dtables/diff_table.go:380]: We return empty row differ async_differ.go when we call `NewRowDiffer`. We check to primarily throw a warning for users. The emptyRowDiffer is useful in reducing the conditional logic checks of schema changes since it just returns an empty diff. Cc. dolt constraints verify

File: go/libraries/doltcore/sqle/tables.go (hunk 7 above)
Comment (VinaiRachakonda) [reply go/libraries/doltcore/sqle/tables.go:1725]: 1. Agree with that refactor
2. Not sure what we're looking for in the integration test. Is it to better assert contents in the index?

File: integration-tests/bats/schema-changes.bats (hunk 8 above)
Comment (VinaiRachakonda) [reply integration-tests/bats/schema-changes.bats:502]: Fair enough haha

File: integration-tests/bats/schema-changes.bats (hunk 9 above)
Comment (VinaiRachakonda) [reply integration-tests/bats/schema-changes.bats:257]: I actually prefer less file clutter in our bats directories. This is technically another type of schema-change.

File: go/libraries/doltcore/schema/column.go (hunk 3 above)
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/schema/column.go:153]: We should adjust the super schema logic instead of changing this. changing this here could have unexpected consequences

File: go/libraries/doltcore/sqle/dtables/diff_table.go (hunk 4 above)
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/sqle/dtables/diff_table.go:380]: My reading of this is that `isDiffablePartition()` will prevent us from ever creating an `async_differ` for two roots that can't be diffed due to incompatible PK sets. Is this not true? 
 We definitely want to catch these cases earlier, higher up the stack

File: integration-tests/bats/schema-changes.bats (hunk 9 above)
Comment (andrew-wm-arthur) [reply integration-tests/bats/schema-changes.bats:257]: Aesthetically, it's subjective, but we've started parallelizing BATS by passing each file to to its own github action. It makes CI quicker to have smaller files

File: go/libraries/doltcore/sqle/tables.go (hunk 7 above)
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/sqle/tables.go:1725]: I'll sync in person

File: integration-tests/bats/primary-key-changes.bats [hunk 10]
```diff
@@ -0,0 +1,539 @@
+#!/usr/bin/env bats
```
Comment (andrew-wm-arthur) [inline integration-tests/bats/primary-key-changes.bats:1]: need to add a test here for failing to add a PK because the column has duplicates

**Review by andrew-wm-arthur — APPROVED** [review]
LGTM
